pkg/validation/utils.go

```diff
@@ -7,6 +7,12 @@ import (
 	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
 )
 
+const (
+	ImportSourceAttribute   = "library.devfile.io/imported-from"
+	ParentOverrideAttribute = "library.devfile.io/parent-override-from"
+	PluginOverrideAttribute = "library.devfile.io/plugin-override-from"
```

Comment:
I was thinking for having the  logic of injecting attributes in api repo before merge/override. However, the attributes need import reference information, the override/merge function has no awareness of those information. In addition, the attributes are used by library only, DWO has their own attributes to check for imported elements. also the generic api should not add additional attributes.  So those can only be added in library. 

Also was thinking if we can check the attributes in library. However, api validator only returns `error`, and concatenated errors for each element, does not have that element information returned

Reply:
but my concern is DWO is doing their own thing and we're expecting odo/ODC are doing their own thing via the devfile/library. Also if DWO is doing their own attribute injection for imports, they are doing unncessary checks when calling validation logic in devfile/api (though it returns the org validation err but they still are being forced to check for these attributes when DWO calls devfile/api validation)

I'm not sure if we want to go this approach since most of the DWO team are the merge approvers of devfile/api. We should talk with them first and see, wdyt?

Reply:
I don't mind having `addSourceAttributesForTemplateSpecContent` under api repo. But if that's the case, then the attribute key should be `api.devfile.io/imported-from`?

Reply:
@amisevsk any thoughts on having the import attribute function stored  under api repo?

Reply:
Hm, this is kinda tricky. I haven't had time to read this PR closely yet (hope to get to it next week!), but I think the cleanest solution (that avoids a circular dependency) would be

1. The function to add source attributes should be in `devfile/library` since, as you say, the context from the resolving process is needed (i.e. what was the import reference for this plugin?)
2. I think the attribute variables should be renamed to `api.devfile.io/imported-from`, etc, and canonically live in `devfile/api`. The `devfile/library` repo should import the variables from here (i.e. [this block](https://github.com/devfile/library/blob/ec1e90d781a1fded63f88084b44e6bf519a9e0ea/pkg/devfile/parser/sourceAttribute.go#L9-L13) should be removed)
3. The annotations stored in this repo should be documented as "here's what you use as an annotation to show where something came from" (but worded nicely).

Basically, `devfile/api` should define how resolved imports are marked (what attribute to assign to a component to signify it came from a specific plugin/parent), and leave the task of actually applying those annotations to implementing libraries. Then whoever writes code that processes devfiles can implement adding this annotation and have it show up from `validate` calls.

Reply:
Also, I'm not sure how `library.devfile.io/parent-override-from` is to be used -- the reason we needed `imported-from` is because it's hard/impossible to know which parent/plugin is responsible for importing a component. Once you can match elements from the flattened devfile to the original devfile, overrides don't need to be marked IMO.

Reply:
Considering for nested parents (parent has another parent reference) situation. If you have multiple parent reference, and each parent will have different overrides. for example: main devfile -> parent 1 -> parent 2 -> parent 3. 
if you only have `import-from` which references parent2, but the actual validation error is caused by the override in parent1; it will be difficult for user to trace the actual error.  `parent-override-from` will be helpful in this case. 
user will see that element is `imported from parent2, overridden in parent1`